backend/src/main.ts

```diff
@@ -1,27 +1,49 @@
-import * as dotenv from "dotenv";
-dotenv.config();
-
 import { createServer } from "http";
-import { Server } from "socket.io";
+import { Server, Socket } from "socket.io";
+import { shareReplay, Subject, take, timeout, } from "rxjs";
+
+import { CoffeeCall, Command, TimedCommand } from "./models";
+import { getAllowedCorsOrigins, getServerListeningPort } from "./environment";
+import { convertCommandsToCalls } from "./message-pipeline";
+
+type VoidListener<T> = (event: T) => void;
+type CoffeeCallerEventsMap = { coffee: VoidListener<TimedCommand>, coffeeCalls: VoidListener<CoffeeCall> };
 
 const httpServer = createServer();
-const io = new Server(httpServer, {
+const io = new Server<CoffeeCallerEventsMap>(httpServer, {
     cors: {
-        origin: process.env.CORS_ALLOWED_ORIGIN || 'http://localhost:3000',
+        origin: getAllowedCorsOrigins(),
         methods: ["GET", "POST"]
     }
 });
 
+const messageSubject = new Subject<Command>();
+const processedCoffeeCalls = messageSubject.pipe(convertCommandsToCalls, shareReplay(1));
+
+function broadcastCommand(command: Command) {
+    const timed = { ...command, broadcastAt: new Date() };
+    console.log('→ broadcast command msg:', timed);
+    io.sockets.emit("coffee", timed);
+}
```

Comment:
With the introduction of aggregated calls in the BE we could basically just remove this.  
Besides code reduction, it'll also make it easier to analyze the protocol by observing the messages sent back and forth (less noise on the line).